src/bin/sol-fbp-generator/main.c

```diff
@@ -755,6 +756,116 @@ collect_context_info(struct generate_context *ctx, struct fbp_data *data)
     return true;
 }
 
+static bool
+generate_memory_map_struct(const char *json_file, int *elements)
+{
+    struct sol_file_reader *fr = NULL;
+    struct sol_json_scanner scanner, entries;
+    struct sol_str_slice contents;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+    uint32_t version;
+
+    *elements = 0;
+
+    fr = sol_file_reader_open(json_file);
+    if (!fr) {
+        SOL_ERR("Couldn't open json file '%s': %s\n", json_file, sol_util_strerrora(errno));
+        return false;
+    }
+
+    contents = sol_file_reader_get_all(fr);
+    sol_json_scanner_init(&scanner, contents.data, contents.len);
+
+    SOL_JSON_SCANNER_ARRAY_LOOP (&scanner, &token, SOL_JSON_TYPE_OBJECT_START, reason) {
+        char *path;
+
+        out("static const struct sol_memmap_map _memmap%d = {\n", (*elements)++);
+
+        SOL_JSON_SCANNER_OBJECT_LOOP_NEST (&scanner, &token, &key, &value, reason) {
+
+            if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "path")) {
+                sol_json_token_remove_quotes(&value);
+                path = strndupa(value.start,
+                    sol_json_token_get_size(&value));
+                if (!path) {
+                    SOL_WRN("Couldn't get map path");
+                    goto error;
+                }
+            } else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "version")) {
+                if (sol_json_token_get_uint32(&value, &version) < 0) {
+                    SOL_WRN("Couldn't get memory map version");
+                    goto error;
+                }
+            } else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "entries")) {
+                sol_json_scanner_init_from_token(&entries, &value);
+
+                out("    .entries = {\n");
+
+                SOL_JSON_SCANNER_ARRAY_LOOP (&entries, &token, SOL_JSON_TYPE_OBJECT_START, reason) {
+                    uint32_t size = 0, offset = 0, bit_offset = 0;
+                    bool is_bool = false;
+                    char *name;
+
+                    SOL_JSON_SCANNER_OBJECT_LOOP_NEST (&entries, &token, &key, &value, reason) {
+                        if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "name")) {
+                            sol_json_token_remove_quotes(&value);
+                            name = strndupa(value.start,
+                                sol_json_token_get_size(&value));
```

Comment:
you don't need to strdupa() just to print it. Use `%.*s`, that way you can provide 2 arguments and limit the string to a number of bytes. If you convert it to a `sol_str_slice` we even have macros to help with that.